projects/distributed-tracing/src/shared/services/metadata/metadata.service.ts

```diff
@@ -116,6 +92,75 @@ export class MetadataService {
     return !isEmpty(attribute.displayName) ? attribute.displayName : attribute.name;
   }
 
+  /**
+   * This is used for adding units to metric restults.
+   * @param rawResult raw result from server
+   * @param specifications all request specifications
+   * @param scope scope of attributes
+   */
+  public buildSpecificationResultWithUnits(
+    rawResult: Dictionary<unknown>,
+    specifications: Specification[],
+    scope: string
+  ): Observable<SpecificationResult[]> {
+    return forkJoinSafeEmpty(
+      specifications.map(spec => {
+        const alias = spec.resultAlias();
+        const data = spec.extractFromServerData(rawResult);
+
+        if (isMetricSpecification(spec) && isMetricAggregation(data)) {
+          return this.resultUnits(scope, spec).pipe(
+            map(units => ({
+              alias: alias,
+              data: { units: units, ...(data as object) }
+            }))
+          );
+        }
+
+        return of({
+          alias: alias,
```

Comment:
So having alias in the result ass far as I can see is just a way of mapping it back to the specific specification that was the source of the mapped data - so can we provide a Map<Specification, SpecficationResult> instead to be more explicit?

Reply:
Although it would make it more readable, I feel I would be complicating the code.  Here is how it would look

```
  public buildSpecificationResultWithUnits(
    rawResult: Dictionary<unknown>,
    specifications: Specification[],
    scope: string
  ): Observable<Map<Specification, SpecificationResult<unknown>>> {
    return forkJoinSafeEmpty(
      specifications.map(spec => {
        const alias = spec.resultAlias();
        const data = spec.extractFromServerData(rawResult);

        if (isMetricSpecification(spec) && isMetricAggregation(data)) {
          return this.resultUnits(scope, spec).pipe(
            map(
              units =>
                [
                  spec,
                  {
                    alias: alias,
                    data: { units: units, ...(data as object) }
                  } as SpecificationResult<unknown>
                ] as [Specification, SpecificationResult<unknown>]
            )
          );
        }

        return of([
          spec,
          {
            alias: alias,
            data: data as SpecificationResult<unknown>
          }
        ] as [Specification, SpecificationResult<unknown>]);
      })
    ).pipe(map(results => new Map<Specification, SpecificationResult<unknown>>(results)));
  }
```

Reply:
On the caller side, the specification in the map would be used. We just iterate over key value, only use value to add it in the final result object. So I am not convinced this would be super helpful. Thoughts?

Reply:
My mistake on the example type - I meant to replace `SpecificationResult` entirely since the map removes need for an alias - so something like `Map<Specification, unknown>`. We do iterate the final value, but maps are still iterable and it's keeping the result matching the input, rather than this implementation being tied to how the caller's using it.

Reply:
Yeah. This makes sense

Reply:
It conflicts with the other comment though :). I would prefer this over the new types.

Reply:
Kind of conflicts, I just got lazy with the unknown - the new type comment could still be relevant here though;
```
export type MaybeWithUnits<T> = T extends MetricAggregation ? T & { units : string} : T;
```
then return `Map<Specification, MaybeWithUnits<T>>`

Reply:
I think for this to work, Specification should itself by typed with T. Since `extractFromServerData` returns unknown, I don't think I can map it to a generic T/MaybeWithUnits<T>. 
```
export interface Specification {
  resultAlias(): string;
  name: string;
  displayName?: string;
  asGraphQlSelections(): GraphQlSelection | GraphQlSelection[];
  asGraphQlOrderByFragment(): GraphQlSortWithoutDirection;
  extractFromServerData(resultContainer: Dictionary<unknown>): unknown;
}
```

Reply:
Spec probably should be typed, but I think the two changes are independent - it's still on the caller to decide what the type of the result is since an array can contain specs with different shapes (hence my comment on the other PR about getting a standard data wrapper object) - this change is more about the returned type giving the caller an indication of what's happening. It's really in the same vane as that comment, I think of it as more of a proto-wrapper object, that we could update in the future. For example, if an entity/span/trace became a dictionary of `MaybeWithUnits` it could still be used as is today, and when we eventually started putting units on all things not just metric aggregations, that would be reflected in the type system vs be transparent in an unknown type. We can hold off on that now and address that as a stand alone effort though.

Reply:
okay. Will address this separately. Will open a tech debt today.